File: services/server/src/protocol/protocol.py
"""
HEADER (3 bytes)
  tipo:          1 byte   (pocos valores)
  largo payload: 2 bytes

PAYLOAD BATCH (variable, tamaño = largo payload)
  agency:        1 byte   
  bets:          repetido, cada uno:
    largo nombre:  1 byte
    nombre:        N bytes   (UTF-8, variable)
    largo apellido:1 byte
    apellido:      M bytes   (UTF-8, variable)
    documento:     4 bytes
    cumpleaños:    4 bytes  (AAAAMMDD como un solo entero)
    number:        4 bytes

header type:
1 = BET       (cliente → servidor, una apuesta) 
4 = END       (cliente → servidor, servidor → cliente, "ya mandé todo")
2 = WINNER   (servidor → cliente, un winner)
3 = ACK   (cliente → servidor, servidor → cliente, confirmación de éxito, sin payload)
5 = BATCH  (cliente → servidor, una o más apuestas concatenadas)

"""
from dataclasses import dataclass

import safe_socket

BET = 1
WINNER = 2
ACK = 3
END = 4
BATCH = 5
_HEADER_SIZE = 3
_MAX_UINT16 = 0xFFFF  # límite de cualquier campo de 2 bytes del protocolo (largo de payload, contadores)
_MAX_PAYLOAD_SIZE = _MAX_UINT16
_MAX_FIELD_LENGTH = 0xFF


@dataclass
class BetMessage:
    nombre: str
    apellido: str
    documento: int
    cumpleanos: str
    number: int

    @staticmethod
    def decode(payload, pos: int = 0) -> tuple['BetMessage', int]:
        """"decodifica una apuesta a partir de pos, y devuelve la pos
        siguiente, para poder decodificar varias apuestas concatenadas 
        dentro de un mismo payload batch"""
        nombre, pos = _read_fixed_field(payload, pos)
        apellido, pos = _read_fixed_field(payload, pos)
        documento = int.from_bytes(payload[pos:pos+4], 'big')
        pos+=4
        cumpleanos = _decode_birthdate(payload[pos:pos+4])
        pos+=4
        number = int.from_bytes(payload[pos:pos+4], 'big')
        pos+=4
        return BetMessage(nombre, apellido, documento, cumpleanos, number), pos

@dataclass
class WinnerMessage:
    nombre: str
    apellido: str
    documento: int
    cumpleanos: str
    number: int

    def encode_payload(self) -> bytes:
        payload = (
            _encode_fixed_field(self.nombre) +
            _encode_fixed_field(self.apellido) +
            self.documento.to_bytes(4, 'big') +
            _encode_birthdate(self.cumpleanos) +
            self.number.to_bytes(4, 'big')
        )
        return payload


def read_message(sock) -> tuple[int, tuple[int, list[BetMessage]] | None]:
    """lee un mensje completo del socket, header + paylaod exacto, 
    y devuevle su tipo con los datos decodificados"""
    header = safe_socket.recv_all(sock, _HEADER_SIZE)
    tipo = int.from_bytes(header[0:1], 'big')
    largo_payload = int.from_bytes(header[1:3], 'big')
    payload = safe_socket.recv_all(sock, largo_payload)
    return tipo, decode_message(tipo, payload)

def read_expected(sock, expected_tipo) -> tuple[bool, tuple[int, list[BetMessage]] | None]:
    """espera leer un mensaje de un tipo puntual, salvo que llege END que 
    implica el fin de la comnicacion. 
    Devuelve (true, none) si fue END o (false, mensaje) si fue el tipo esperado 
    cualquier otro es un error"""
    tipo, data = read_message(sock)
    if tipo == END:
        return True, None
    if tipo != expected_tipo:
        raise ValueError(f"mensaje inesperado, se esperaba {expected_tipo} o END: tipo={tipo}")
    return False, data

def encode_end() -> bytes:
    return encode_message(END, b'')

def encode_ack() -> bytes:
    return encode_message(ACK, b'')

def encode_winner(winner: WinnerMessage) -> bytes:
    return encode_message(WINNER, winner.encode_payload())

def encode_message(tipo: int, payload: bytes) -> bytes:
    """arma el mensaje completo validando que el largo del payload
    no supere el maximo"""
    if len(payload) > _MAX_PAYLOAD_SIZE:
        raise ValueError(f"payload de {len(payload)} bytes excede el máximo de {_MAX_PAYLOAD_SIZE}")
    return tipo.to_bytes(1, 'big') + len(payload).to_bytes(2, 'big') + payload

def _encode_fixed_field(value: str) -> bytes:
    """"codifica un campo de largo variable con 1 byte de largo + el contenido"""
    value_bytes = value.encode('utf-8')
    if len(value_bytes) > _MAX_FIELD_LENGTH:
        raise ValueError(f"campo de {len(value_bytes)} bytes excede el máximo de {_MAX_FIELD_LENGTH}")
    return len(value_bytes).to_bytes(1, 'big') + value_bytes

def _encode_birthdate(cumpleanos) -> bytes:
    " de AAAA-MM-DD a AAAAMMDD bytes"
    cumpleanos_int = int(cumpleanos.replace("-", ""))
    return cumpleanos_int.to_bytes(4, 'big')

def decode_batch(payload) -> tuple[int, list[BetMessage]]:
    """decodifica el payload de un mensaje BATCH: el agency id, seguido
    de todoas las BETs concatenadas"""
    agency = int.from_bytes(payload[0:1], 'big')
    pos = 1 
    bets = []
    while pos < len(payload):
        bet, pos = BetMessage.decode(payload, pos)
        bets.append(bet)
    return agency, bets

def _decode_birthdate(birthday_bytes) -> str:
    "de AAAAMMDD a AAAA-MM-DD"
    birthday = int.from_bytes(birthday_bytes, 'big')
    birthday_str = str(birthday)
    if len(birthday_str) != 8:
        raise ValueError(f"fecha de nacimiento inválida: se esperaban 8 dígitos (AAAAMMDD), se recibió {birthday_str!r}")
    year = birthday_str[:4]
    month = birthday_str[4:6]
    day = birthday_str[6:8]
    return f"{year}-{month}-{day}"

def decode_message(tipo, payload) -> tuple[int, list[BetMessage]] | None:
    if tipo == BATCH:
        return decode_batch(payload)
    elif tipo == ACK:
        return None
    elif tipo == END:
        return None
    else:
        raise ValueError(f"Tipo de mensaje desconocido: {tipo}")

def _read_fixed_field(payload: bytes, pos: int) -> tuple[str, int]:
    """lee el byte de largo y despues esa cantidad de bytes de contenido"""
    length = payload[pos]
    pos+=1
    value_bytes = payload[pos:pos+length]
    pos+=length
    return value_bytes.decode('utf-8'), pos

File: services/client/src/protocol/protocol.go
package protocol

/* HEADER (3 bytes)
  tipo:          1 byte
  largo payload: 2 bytes

PAYLOAD BATCH (variable, tamaño = largo payload)
  agency:        1 byte
  bets:          repetido, cada uno:
    largo nombre:  1 byte
    nombre:        N bytes   (UTF-8, variable)
    largo apellido:1 byte
    apellido:      M bytes   (UTF-8, variable)
    documento:     4 bytes
    cumpleaños:    4 bytes  (AAAAMMDD como un solo entero)
    number:        4 bytes

header type:
1 = BET       (cliente → servidor, una apuesta)
4 = END       (cliente → servidor, servidor → cliente, "ya mande todo")
2 = WINNER   (servidor → cliente, un winner)
3 = ACK   (cliente → servidor, servidor → cliente, confirmación de exito, sin payload)
5 = BATCH ( client -> servidor)
*/

import (
	"encoding/binary"
	"fmt"
	"net"
	"strconv"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const Winner = 2
const Ack = 3
const End = 4
const Batch = 5
const headerSize = 3
const MaxPayloadSize = 0xFFFF
const MaxFieldLength = 0xFF

type WinnerMessage struct {
	Name      string
	Lastname  string
	Document  uint32
	Birthdate string
	Number    uint32
}

func ReadMessage(sock net.Conn) (int, []byte, error) {
	header, err := safe_socket.RecvAll(sock, headerSize)

	if err != nil {
		return 0, nil, err
	}

	tipo := int(header[0])
	largo_payload := int(binary.BigEndian.Uint16(header[1:3]))
	payload, err := safe_socket.RecvAll(sock, largo_payload)

	if err != nil {
		return 0, nil, err
	}

	return tipo, payload, nil
}

func EncodeEnd() []byte {
	return newMessage(End, 0)
}

func EncodeAck() []byte {
	return newMessage(Ack, 0)
}

// codifica una apuesta y la agrega al final del buf
// se usa para acumular apuestas dentro de un batch
func AppendBet(buf []byte, name, lastname, birthdate []byte, document, number uint32) ([]byte, error) {
	buf, err := appendFixedField(buf, name)

	if err != nil {
		return nil, fmt.Errorf("name: %w", err)
	}

	buf, err = appendFixedField(buf, lastname)

	if err != nil {
		return nil, fmt.Errorf("lastname: %w", err)
	}

	birthdateValue, err := encodeBirthdate(birthdate)
	if err != nil {
		return nil, fmt.Errorf("birthdate: %w", err)
	}

	buf = binary.BigEndian.AppendUint32(buf, document)
	buf = binary.BigEndian.AppendUint32(buf, birthdateValue)
	buf = binary.BigEndian.AppendUint32(buf, number)
	return buf, nil
}

// una vez agregado el agency y payload se calcula y guarda
// el largo del payload en el header
func AppendBatch(buf []byte, agency byte, payload []byte) ([]byte, error) {
	headerPos := len(buf)
	buf = append(buf, Batch, 0, 0)
	payloadStart := len(buf)

	buf = append(buf, agency)
	buf = append(buf, payload...)

	payloadSize := len(buf) - payloadStart

	if payloadSize > MaxPayloadSize {
		return nil, fmt.Errorf("batch payload de %d bytes excede el maximo de %d", payloadSize, MaxPayloadSize)
	}

	binary.BigEndian.PutUint16(buf[headerPos+1:headerPos+3], uint16(payloadSize))
	return buf, nil
}

func newMessage(tipo byte, payloadSize int) []byte {
	message := make([]byte, headerSize+payloadSize)
	message[0] = tipo
	binary.BigEndian.PutUint16(message[1:3], uint16(payloadSize))
	return message
}

// saca los guiones de "AA-MM-DD" y valida que queden 8 numeros
func encodeBirthdate(birthdate []byte) (uint32, error) {
	var numbers [8]byte
	n := 0
	for _, b := range birthdate {
		if b == '-' {
			continue
		}
		if n >= len(numbers) {
			return 0, fmt.Errorf("fecha de nacimiento invalida: se esperaban 8 digitos (AAAAMMDD), se recibio %q", birthdate)
		}
		numbers[n] = b
		n++
	}
	if n != len(numbers) {
		return 0, fmt.Errorf("fecha de nacimiento invalida: se esperaban 8 digitos (AAAAMMDD), se recibio %q", birthdate)
	}

	value, err := strconv.ParseUint(string(numbers[:]), 10, 32)
	if err != nil {
		return 0, fmt.Errorf("fecha de nacimiento invalida: %w", err)
	}
	return uint32(value), nil
}

// decodificamos el payload de un mensaje winner
// pos a pos, como hay campos de largos dinamicos que estan
// precedidos de un largo de tamaño fijo que lo define
func DecodeWinner(payload []byte) (WinnerMessage, error) {
	name, pos := readFixedField(payload, 0)
	lastname, pos := readFixedField(payload, pos)

	documentBytes := payload[pos : pos+4]
	pos += 4
	document := binary.BigEndian.Uint32(documentBytes)

	birthdateBytes := payload[pos : pos+4]
	pos += 4
	birthdate, err := decodeBirthdate(birthdateBytes)
	if err != nil {
		return WinnerMessage{}, fmt.Errorf("birthdate: %w", err)
	}

	numberBytes := payload[pos : pos+4]
	number := binary.BigEndian.Uint32(numberBytes)

	return WinnerMessage{name, lastname, document, birthdate, number}, nil
}

// inverso de encodeBirthday
func decodeBirthdate(birthdayBytes []byte) (string, error) {
	value := binary.BigEndian.Uint32(birthdayBytes)
	str := strconv.FormatUint(uint64(value), 10)
	if len(str) != 8 {
		return "", fmt.Errorf("fecha de nacimiento invalida: se esperaban 8 digitos (AAAAMMDD), se recibio %q", str)
	}
	return str[0:4] + "-" + str[4:6] + "-" + str[6:8], nil
}

// para los campos de largo variable, codificamos el
// largo definido primero -> el prefijo del byte limita el
// largo posible del campo variable
func appendFixedField(buf []byte, value []byte) ([]byte, error) {
	if len(value) > MaxFieldLength {
		return nil, fmt.Errorf("campo de  %d bytes excede el maximo de %d", len(value), MaxFieldLength)
	}
	buf = append(buf, byte(len(value)))
	buf = append(buf, value...)
	return buf, nil
}

// apra los campos de largo variable, leemos primero el largo
// y despues es cantidad de contenido
func readFixedField(payload []byte, pos int) (string, int) {
	length := int(payload[pos])
	pos++

	value := payload[pos : pos+length]
	pos += length
	return string(value), pos
}
